### PR TITLE
refactor manual function definitions

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43718 make jvp use forward mode by default
* #43717 Add gradcheck for forward grad
* #43716 Add many forward grad
* #43715 Basic auto formula for forward grad
* #43714 make sure SavedVariable propagate fw_grad properly
* #43713 Add forward grad mode in python and c++
* #43712 Add fw_grad field and accessors
* **#43711 refactor manual function definitions**

this makes them available in forward if needed

No change to the file content, just a copy-paste.

Differential Revision: [D23454146](https://our.internmc.facebook.com/intern/diff/D23454146)

### CODE REVIEW

File: tools/autograd/templates/ManualFunctions.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <ciso646>
+#include <algorithm>
+#include <numeric>
+#include <functional>
+
+using at::Tensor;
+using at::Scalar;
+using at::IntArrayRef;
+using at::TensorList;
```
Comment (zou3519) [inline tools/autograd/templates/ManualFunctions.h:29]: Is ManualFunctions.h included in another C++ file in this stack of PRs, or is it just used in VariableType.cpp?

The `using at::Tensor;` declarations make Tensor, Scalar, IntArrayRef, and TensorList available to the global namespace. This could be bad if ManualFunctions.h ultimately gets included into e.g. torch.h, so it's usually better for `using blah` to appear in the .cpp files

Comment (albanD) [reply tools/autograd/templates/ManualFunctions.h:29]: No it is not added anywhere else in this stack.
It's not actually a header file as it's just added to the cpp files.

Do you think I should remove the `.h` that can be confusing?

File: tools/autograd/templates/VariableType.cpp [hunk 2]
```diff
@@ -29,8 +29,14 @@
 using namespace at;
 using namespace torch::autograd::generated;
 
+${manual_functions_h}
+
 namespace torch { namespace autograd {
 
+namespace {
+${manual_functions_cpp}
```
Comment (zou3519) [inline tools/autograd/templates/VariableType.cpp:37]: If I'm understanding this correctly, manual_functions.cpp will be compiled into VariableType.cpp AND Functions.cpp.

This would increase the binary size, but probably not by a very significant amount (the compiler hopefully knows how to throw away the parts that don't get used).

What are the functions that you wanted to share between the forward implementations and Functions.cpp? Is it only a few or is it a lot of things? If we were writing straight-up C++ we would define things to be shared in a header and include the header where necessary, but the codegen does give us a lot more freedom

Comment (albanD) [reply tools/autograd/templates/VariableType.cpp:37]: It turns out that most of the forward formulas actually need the backward definition as well.
I went with this version to:
- Keep the anonymous namespace to make sure we don't expose these in other places by mistake
- Assumed that the compiler will just remove everything that's not needed indeed
- Keep the manual Functions definition close to where they are right now so that people don't get lost looking for them.

Comment (zou3519) [reply tools/autograd/templates/VariableType.cpp:37]: Well, this also increases compilation time because the code now gets compiled as a part of N compilation units, where N is the number of VariableType shards we do (5?). Furthermore, modifying something in ManualFunctions.cpp will trigger a recompile of all of the VariableType shards.

I'm still wary of adding new things to the codegen. However, the codegen changes are pretty minimal, so we can make the case that this isn't too crazy of a change.

As an alternative, would something like the following work?
- Add ManualFunctions.h and ManualFunctions.cpp.
- ManualFunctions.h contains declarations for all of the functions in Functions.cpp in a new torch::autograd::generated::details namespace
- ManualFunctions.cpp contains definitions for all of the functions in Functions.cpp in torch::autograd::generated::details namespace
- `#include ManualFunctions.h` into Functions.cpp and VariableType.cpp (and just use `using torch::autograd::generated::details` inside those .cpp files).

This way, we don't expose any symbols in other places, keep the manual functions definition close to where they are now, don't need to modify the codegen, and keep a nice header/body split. The only thing that is tedious is listing declarations in ManualFunction.h (but one can probably write a script to toss out the names of all functions in Functions.cpp)

Comment (zou3519) [inline tools/autograd/templates/VariableType.cpp:37]: What is the reason why we're including ManualFunctions into VariableType? Are the forward mode formulas for e.g. `torch.abs` inlined directly into VariableType::abs ?

Comment (ezyang) [reply tools/autograd/templates/VariableType.cpp:37]: I agree with @zou3519. Even if you don't want to put things in a cpp file and duplicate them, you should simply mark all the functions as `inline` and put them in a header (and don't transclude them manually via codegen). But it is much better to put them in a cpp file so there is only one copy floating around.

Comment (albanD) [reply tools/autograd/templates/VariableType.cpp:37]: Yes after a quick chat with Richard, I will move these out of the codegen code and just include them once.

File: caffe2/CMakeLists.txt [hunk 3]
```diff
@@ -453,6 +453,7 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     list(APPEND TORCH_SRCS
```
Comment (ljk53) [inline caffe2/CMakeLists.txt:453]: Note that mobile build has set `INTERN_DISABLE_AUTOGRAD` at the line above.
These files are supposed not needed for inference-only mobile build.

Comment (albanD) [reply caffe2/CMakeLists.txt:453]: Good point, it seems like `Functions.cpp` is treated differently than the other codegened files and is included in the mobile build!
So I'll add this one file as well.

File: tools/autograd/templates/VariableType.cpp (hunk 2 above)
Comment (albanD) [reply tools/autograd/templates/VariableType.cpp]: Yes it is. The helper code has to be there (as we re-use values from this function) and the call to the formula itself is a one liner (either containing the formula or calling code in aten/ManualFunctions). So these are directly inlined.

File: tools/autograd/templates/Functions.cpp [hunk 4]
```diff
@@ -1,2880 +1,19 @@
...
-#include <algorithm>
-#include <numeric>
-#include <functional>
+#include "torch/csrc/autograd/FunctionsManual.h"
 
 // ${generated_comment}
 
+// The manual function definitions that used to be here are now in torch/csrc/autograd/FunctionsManual.cpp
+// This speeds up re-compilation and allow to share these implementations so that they can be
+// used for forward formulas as well.
```
Comment (zou3519) [inline tools/autograd/templates/Functions.cpp:7]: nit: "forward formulas" -> "forward mode AD formulas", to not get confused with "forward passes"

File: tools/autograd/templates/Functions.h [hunk 5]
```diff
@@ -10,7 +10,7 @@
 #include "torch/csrc/autograd/function.h"
 #include "torch/csrc/autograd/variable.h"
 #include "torch/csrc/autograd/saved_variable.h"
-#include <torch/csrc/WindowsTorchApiMacro.h>
+#include "torch/csrc/WindowsTorchApiMacro.h"
```
Comment (zou3519) [inline tools/autograd/templates/Functions.h:13]: Why the change?

File: tools/build_variables.bzl [hunk 6]
```diff
@@ -42,6 +42,7 @@ def libtorch_generated_sources(gencode_pattern):
     ]] + [
         "torch/csrc/autograd/TraceTypeManual.cpp",
         "torch/csrc/autograd/VariableTypeManual.cpp",
+        "torch/csrc/autograd/FunctionsManual.cpp",
```
Comment (zou3519) [inline tools/build_variables.bzl:45]: I'm confused because none of these three files are autogenerated. But as long as the build work this seems fine

**Review by zou3519 — APPROVED** [review]
LGTM. RIP the git blame data but that's okay

File: tools/autograd/templates/Functions.cpp (hunk 4 above)
Comment (zou3519) [reply tools/autograd/templates/Functions.cpp:7]: (feel free to address this in a follow-up, since we should try to merge this sooner rather than later to avoid merge conflicts)

File: tools/build_variables.bzl (hunk 6 above)
Comment (albanD) [reply tools/build_variables.bzl:45]: They correspond to the functions that are missing from the codegen. So they are at the same place.

File: tools/autograd/templates/Functions.h (hunk 5 above)
Comment (albanD) [reply tools/autograd/templates/Functions.h:13]: remain from previous code move. I'll revert that.
